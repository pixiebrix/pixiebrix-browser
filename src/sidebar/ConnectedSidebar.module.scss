/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

@import "@/themes/colors";

// Due to design system inconsistencies,
// we have to override our primary purple color
// for the PixieBrix theme
.themeColorOverride {
  color: $N800;

  &:hover {
    color: $N300;
  }
}

.themeColor {
  color: var(--theme-primary);

  &:hover {
    color: var(--theme-primary);
    background-color: rgba(0, 0, 0, 0.07);
  }
}

.button {
  display: inline-flex;
  align-items: center;
  text-decoration: none;
}

.logo {
  // see _themes.scss
  height: var(--logo-height);
  padding: 0 1.5rem;
}
