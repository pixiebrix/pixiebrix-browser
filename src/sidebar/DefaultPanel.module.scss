/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

.root {
  p {
    font-size: 16px;
    color: #41335b;
    font-weight: 400;
  }

  a {
    color: #170637;
    font-weight: 500;
    font-size: 16px;
    padding-bottom: 1px;
    border-bottom: 2px solid #5a19da;
    text-decoration: none;
  }

  kbd {
    color: #170637;
    font-weight: 500;
    font-size: 16px;
    background: white;
    border-radius: 4px;
    padding: 4.5px 6px;
    margin: 0 12px;
  }
}

.sidebarRow {
  text-align: center;
  padding: 20px 0;
  margin-top: 28px;
}

.callout {
  color: #170637;
  font-size: 25px;
  font-weight: 700;
}

.sidebarRowWithDivider {
  @extend .sidebarRow;
  border-top: 1px solid #e1dfe5;
}

.tinyCallout {
  @extend .callout;
  font-size: 16px;
}
