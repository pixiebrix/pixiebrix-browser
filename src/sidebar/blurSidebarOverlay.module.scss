/*!
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

.blurOverlay {
  display: flex;
  justify-content: center;
  align-items: flex-start;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(255, 255, 255, 0.5);
  backdrop-filter: blur(2.5px);
  pointer-events: all;
  z-index: 1; // ensure overlay is on top of the sidebar panel content
  > div {
    margin-top: 15vh;
    width: 300px;
  }
}

.modalDialog {
  text-align: center;
  margin: 23px;
  padding: 16px;
  border-radius: 12px;
  background: white;
  border: 1px solid #cfcbd6;
  > div {
    border: 0;
  }
}

.modalHeader {
  display: block;
  padding-bottom: 0;
  background: white;
  border: 0;
}

.modalBody {
  background: white;
  border: 0;
}
