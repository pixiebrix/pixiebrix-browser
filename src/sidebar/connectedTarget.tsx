/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { isMV3 } from "@/mv3/api";
import { expectContext } from "@/utils/expectContext";
import { assertNotNullish } from "@/utils/nullishUtils";
import { once } from "lodash";
import { type TopLevelFrame, getTopLevelFrame } from "webext-messenger";
import { getTabUrl } from "webext-tools";

function getConnectedTabIdMv3(): number {
  expectContext("sidebar");
  const tabId = new URLSearchParams(window.location.search).get("tabId");
  assertNotNullish(
    tabId,
    `No tabId argument was found on this page: ${window.location.href}`,
  );
  return Number(tabId);
}

async function getConnectedTabIdMv2() {
  const { tabId } = await getTopLevelFrame();
  return tabId;
}

export const getConnectedTabId = once(
  isMV3() ? getConnectedTabIdMv3 : getConnectedTabIdMv2,
);
export const getConnectedTarget = isMV3()
  ? (): TopLevelFrame => ({ tabId: getConnectedTabIdMv3(), frameId: 0 })
  : getTopLevelFrame;

export async function getAssociatedTargetUrl(): Promise<string | undefined> {
  return getTabUrl(await getConnectedTarget());
}
