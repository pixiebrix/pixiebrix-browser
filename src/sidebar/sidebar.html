<!--
  ~ Copyright (C) 2022 PixieBrix, Inc.
  ~
  ~ This program is free software: you can redistribute it and/or modify
  ~ it under the terms of the GNU Affero General Public License as published by
  ~ the Free Software Foundation, either version 3 of the License, or
  ~ (at your option) any later version.
  ~
  ~ This program is distributed in the hope that it will be useful,
  ~ but WITHOUT ANY WARRANTY; without even the implied warranty of
  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  ~ GNU Affero General Public License for more details.
  ~
  ~ You should have received a copy of the GNU Affero General Public License
  ~ along with this program.  If not, see <http://www.gnu.org/licenses/>.
  -->

<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <!-- Open all links in new tab, never in the sidebar #851 -->
    <base target="_blank" />
    <title>PixieBrix: Page Panel</title>
    <link rel="stylesheet" href="loadingScreen.css" />
    <link rel="stylesheet" href="sidebar.css" />
  </head>
  <body>
    <div id="container" class="full-height vh-100">
      <div class="global-loading-message">Loading Sidebar ðŸš€</div>
    </div>
    <script src="loadingScreen.js"></script>
    <script src="sidebar.js"></script>
    <!-- Must be at the bottom https://github.com/pixiebrix/pixiebrix-extension/issues/4100 -->
    <link
      href="https://fonts.googleapis.com/css2?family=Ubuntu:wght@300;400;500;700&display=swap"
      rel="stylesheet"
    />
  </body>
</html>
