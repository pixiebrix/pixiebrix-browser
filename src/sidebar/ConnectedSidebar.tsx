/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "./ConnectedSidebar.module.scss";

import React, { Dispatch, useEffect, useMemo } from "react";
import { Button } from "react-bootstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faAngleDoubleRight, faCog } from "@fortawesome/free-solid-svg-icons";
import {
  addListener,
  removeListener,
  SidebarListener,
} from "@/sidebar/protocol";
import DefaultPanel from "@/sidebar/DefaultPanel";
import { useDispatch, useSelector } from "react-redux";
import { ActivatePanelOptions, FormEntry, PanelEntry } from "@/sidebar/types";
import Tabs from "@/sidebar/Tabs";
import sidebarSlice, { SidebarState } from "./sidebarSlice";
import { AnyAction } from "redux";
import { hideSidebar } from "@/contentScript/messenger/api";
import { whoAmI } from "@/background/messenger/api";
import useTheme from "@/hooks/useTheme";
import RequireAuth from "@/auth/RequireAuth";
import LoginPanel from "@/sidebar/LoginPanel";

/**
 * Listeners to update the Sidebar's Redux state upon receiving messages from the contentScript.
 */
function getConnectedListener(dispatch: Dispatch<AnyAction>): SidebarListener {
  return {
    onRenderPanels(panels: PanelEntry[]) {
      dispatch(sidebarSlice.actions.setPanels({ panels }));
    },
    onShowForm(form: FormEntry) {
      dispatch(sidebarSlice.actions.addForm({ form }));
    },
    onHideForm({ nonce }: Partial<FormEntry>) {
      dispatch(sidebarSlice.actions.removeForm(nonce));
    },
    onActivatePanel(options: ActivatePanelOptions) {
      dispatch(sidebarSlice.actions.activatePanel(options));
    },
  };
}

const selectState = ({ sidebar }: { sidebar: SidebarState }) => sidebar;

const ConnectedSidebar: React.VFC = () => {
  const { logo } = useTheme();
  const dispatch = useDispatch();
  const sidebarState = useSelector(selectState);

  const listener: SidebarListener = useMemo(
    () => getConnectedListener(dispatch),
    [dispatch]
  );

  // `effect` will run once on component mount since listener and formsRef don't change on renders
  useEffect(() => {
    addListener(listener);
    return () => {
      // NOTE: we don't need to cancel any outstanding forms on unmount because the FormTransformer is set up to watch
      // for PANEL_HIDING_EVENT. (and the only time this SidebarApp would unmount is if the sidebar was closing)
      removeListener(listener);
    };
  }, [listener]);

  return (
    <div className="full-height">
      <div className="d-flex p-2 justify-content-between align-content-center">
        <Button
          className={styles.button}
          onClick={async () => {
            const sidebar = await whoAmI();
            await hideSidebar({ tabId: sidebar.tab.id });
          }}
          size="sm"
          variant="link"
        >
          <FontAwesomeIcon icon={faAngleDoubleRight} className="fa-lg" />
        </Button>
        <div className="align-self-center">
          <img
            src={logo.regular}
            alt="PixieBrix logo"
            className={styles.logo}
          />
        </div>
        <Button
          href="/options.html"
          target="_blank"
          size="sm"
          variant="link"
          className={styles.button}
        >
          <span>
            Options <FontAwesomeIcon icon={faCog} />
          </span>
        </Button>
      </div>

      <div className="full-height">
        <RequireAuth LoginPage={LoginPanel}>
          {sidebarState.panels?.length || sidebarState.forms?.length ? (
            <Tabs
              {...sidebarState}
              onSelectTab={(eventKey: string) => {
                dispatch(sidebarSlice.actions.selectTab(eventKey));
              }}
            />
          ) : (
            <DefaultPanel />
          )}
        </RequireAuth>
      </div>
    </div>
  );
};

export default ConnectedSidebar;
