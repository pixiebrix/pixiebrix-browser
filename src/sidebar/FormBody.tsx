/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { type FormEntry } from "@/sidebar/types";
import { useAsyncState } from "@/hooks/common";
import Loader from "@/components/Loader";
import { getErrorMessage } from "@/errors/errorHelpers";
import { createFrameSource } from "@/blocks/transformers/ephemeralForm/formTransformer";

type FormBodyProps = {
  form: FormEntry;
};

/**
 * JSON Schema form for embedding in a sidebar tab
 * @param form the form definition and extension metadata
 * @constructor
 */
const FormBody: React.FunctionComponent<FormBodyProps> = ({ form }) => {
  const [sourceURL, isLoading, error] = useAsyncState(
    async () => createFrameSource(form.nonce, "panel"),
    [form.nonce]
  );

  if (isLoading) {
    return (
      <div>
        <Loader />
      </div>
    );
  }

  if (error) {
    return (
      <div className="text-danger">
        Error getting information for form: {getErrorMessage(error)}
      </div>
    );
  }

  return (
    <iframe
      title={form.nonce}
      height="100%"
      width="100%"
      src={sourceURL.toString()}
      style={{ border: "none" }}
      allowFullScreen={false}
    />
  );
};

export default FormBody;
