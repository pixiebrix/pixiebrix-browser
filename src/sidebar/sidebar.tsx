/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// Keep in order so precedence is preserved
import "@/vendors/theme/app/app.scss";
import "@/vendors/overrides.scss";
import "@/utils/layout.scss";
import "./sidebar.scss";

import "@/extensionContext";

import registerMessenger from "@/sidebar/messenger/registration";
import App from "@/sidebar/SidebarApp";
import ReactDOM from "react-dom";
import React from "react";
import registerBuiltinBlocks from "@/blocks/registerBuiltinBlocks";
import registerContribBlocks from "@/contrib/registerContribBlocks";
import { initToaster } from "@/utils/notify";

registerMessenger();
registerContribBlocks();
registerBuiltinBlocks();
initToaster();

ReactDOM.render(<App />, document.querySelector("#container"));
