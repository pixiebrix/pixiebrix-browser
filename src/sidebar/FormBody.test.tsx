/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render } from "@/sidebar/testHelpers";
import { formEntryFactory } from "@/testUtils/factories";
import FormBody from "@/sidebar/FormBody";
import { waitForEffect } from "@/testUtils/testHelpers";

jest.mock("@/blocks/transformers/ephemeralForm/formTransformer", () => ({
  createFrameSource: jest.fn(() => new URL("https://www.testUrl.com")),
}));

describe("FormBody", () => {
  it("renders successfully", async () => {
    const form = formEntryFactory();
    const rendered = render(<FormBody form={form} />);
    await waitForEffect();
    expect(rendered.asFragment()).toMatchSnapshot();
  });
});
