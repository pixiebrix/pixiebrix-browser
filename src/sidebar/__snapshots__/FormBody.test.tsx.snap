// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormBody renders successfully 1`] = `
<DocumentFragment>
  <iframe
    height="100%"
    src="https://www.testurl.com/"
    title="00000001-0000-4000-A000-000000000000"
    width="100%"
  />
</DocumentFragment>
`;
